### ISSUE TITLE
 Running on Amazon Lambda

### ISSUE THREAD

**Issue by JayVem** [coolwanglu/pdf2htmlEX#694]
Is it possible to run pdf2htmlEx on Amazon Lambda? Amazon uses its own Amazon Linux on compute instances that use Lambda.

**Comment by davidhedley** [comment]
Unlikely ever to happen. Lambda supported only Node.js, Java and Python. pdf2htmlEX is C++ and relies on a whole bunch of supporting libraries.

**Comment by tylerjryan** [comment]
Lambda supported languages only restrict the language that function code can be written in, not the code that can be executed. For example, I can write a python Lambda function that calls compiled, executable c++ and some linux command line utility bundled with my function. AWS wrote a blog post about running arbitrary executables on Lambda last year:

https://aws.amazon.com/blogs/compute/running-executables-in-aws-lambda/

In theory, though it may require some work, the c++ can be compiled into executables (ahead of time using gcc, like this http://stackoverflow.com/questions/1140995/build-an-linux-executable-using-gcc) that can be called from a "supported" language like python or node.js, and the necessary libraries can be bundled into the Lambda function zip file. 

So to my understanding, it may require some rework of how pdf2htmlEX runs, but is possible.

**Comment by JayVem** [comment]
Yes, I got it working by compiling it on amazon linux and packing the executable with the lambda.

**Comment by tylerjryan** [comment]
@JayVem That's great! Would you be willing to share your function with the community?

**Comment by JayVem** [comment]
the function itself is very simple - it just invokes pdf2htmlex as an external system process using nodejs. The difficult part was compiling pdf2htmlEX on Amazon Linux.

**Comment by fasiha** [comment]
Emscripten?

**Comment by devforgalaxy** [comment]
Any progress or alternative solution?

**Comment by careerlister** [comment]
@JayVem Can you point me in the direction of instructions or provide instructions on how you were able to compile pdf2htmlex and the dependencies for Lambda?   That would be greatly appreciated as we are attempting do so without success.

**Comment by ardcore** [comment]
@JayVem I'd also be interested -- I got to a point where I managed to compile pdf2htmlEX on Amazon Linux, but not a static binary, and I think static binary will be required here. Are you able to share some insights/the binary itself, or is it proprietary?

**Comment by dengelke** [comment]
@careerlister @ardcore I've managed to get it working on lambda (finally). Approach I eventually took was:

1. Building all dependent libraries and pdf2htmlEX from source on a [lambci/lambda:build-nodejs6.10](https://github.com/lambci/docker-lambda) docker build image locally to deal with differences between the lambda linux environment and the host OS (in my case OSX). Initially tried on a remote Amazon Linux EC2 instance but the output didn't work on Lambda.
2. After this move the required libraries & binaries to a lambda function on your host machine.
3. Run on a image that replicates your desired lambda environment, my test script basically was:
```docker run -v \"$PWD\":/var/task lambci/lambda:nodejs6.10 index.handler '{"some":"event"}'``` using a [child process](https://nodejs.org/docs/latest-v6.x/api/child_process.html) inside node.js to call pdf2htmlEX from node.
4. Adjust/repeat steps 1, 2 & 3 until the function works on the local docker image.
5. Deploy to lambda ensuring your total deployment size unzipped is less then 250mb see [here](https://hackernoon.com/exploring-the-aws-lambda-deployment-limits-9a8384b0bec3) for an explanation. Due to using the docker image locally to test, I didn't have any issues migrating over with different operating environments.

**Comment by gutitrombotto** [comment]
@JayVem Can you tell me how did you installed pdf2htmlex on amazon linux? I can't install it. I have ubuntu 18 distro
